### PR TITLE
Import extension settings

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/8847

- Import extension settings 
- Added tests to extensions importer

https://user-images.githubusercontent.com/2965009/188576413-a71aa528-0693-463e-b1a1-44b95110eaa3.mp4

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [x] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
- Install Metamask to Google Chrome
- Login/create wallet in MM
- Import extensions from Chrome to Brave
- Check that MM has same state in Brave as it was befor in Chrome

### CODE REVIEW

File: browser/importer/extensions_import_helpers.cc [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+void ImportStorages(base::FilePath source_profile,
+                    base::FilePath target_profile,
+                    std::vector<std::string> extensions_ids) {
+  DCHECK(
+      extensions::GetExtensionFileTaskRunner()->RunsTasksInCurrentSequence());
+  auto ids_with_settings =
+      GetChromeExtensionsSettingsList(source_profile, extensions_ids);
+
+  auto source_storages = CreateStorages(
```
Comment (simonhong) [inline browser/importer/extensions_import_helpers.cc:66]: nit: `const auto`

File: browser/importer/extensions_import_helpers.cc [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+}  // namespace
+
+void ImportStorages(base::FilePath source_profile,
+                    base::FilePath target_profile,
+                    std::vector<std::string> extensions_ids) {
+  DCHECK(
+      extensions::GetExtensionFileTaskRunner()->RunsTasksInCurrentSequence());
+  auto ids_with_settings =
```
Comment (simonhong) [inline browser/importer/extensions_import_helpers.cc:63]: nit: `const auto`

File: browser/importer/extensions_import_helpers.cc [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+#include "components/value_store/value_store_factory.h"
+#include "components/value_store/value_store_factory_impl.h"
+#include "extensions/browser/api/storage/value_store_util.h"
+#include "extensions/browser/extension_file_task_runner.h"
+#include "extensions/common/constants.h"
+
+namespace brave {
+namespace {
+using ExtensionStorageMap =
+    std::unordered_map<std::string, std::unique_ptr<value_store::ValueStore>>;
```
Comment (simonhong) [inline browser/importer/extensions_import_helpers.cc:25]: maybe `base::flat_map` is preferred over `std::unordered_map`?

File: browser/importer/extensions_import_helpers.cc [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  DCHECK(
+      extensions::GetExtensionFileTaskRunner()->RunsTasksInCurrentSequence());
+  auto ids_with_settings =
+      GetChromeExtensionsSettingsList(source_profile, extensions_ids);
+
+  auto source_storages = CreateStorages(
+      source_profile, ids_with_settings,
+      base::MakeRefCounted<value_store::ValueStoreFactoryImpl>(source_profile));
+
+  auto target_storages = CreateStorages(
```
Comment (simonhong) [inline browser/importer/extensions_import_helpers.cc:70]: what happens if user already has some extensions before importing?
Maybe it's overwritten with chrome's one.
IMO, we should respect user's existing settings.

File: browser/importer/extensions_import_helpers.cc (hunk 1 above)
Comment (spylogsster) [reply browser/importer/extensions_import_helpers.cc]: done, thanks

File: browser/importer/extensions_import_helpers.cc (hunk 2 above)
Comment (spylogsster) [reply browser/importer/extensions_import_helpers.cc]: done

File: browser/importer/extensions_import_helpers.cc (hunk 3 above)
Comment (spylogsster) [reply browser/importer/extensions_import_helpers.cc]: done

File: browser/importer/extensions_import_helpers.cc (hunk 4 above)
Comment (spylogsster) [reply browser/importer/extensions_import_helpers.cc:74]: sure, nice catch, added it in the last commit, thanks!

File: browser/importer/extensions_import_helpers.cc [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+#include "base/files/file_util.h"
+#include "base/logging.h"
+#include "base/task/thread_pool.h"
+#include "brave/common/importer/importer_constants.h"
+#include "components/value_store/value_store.h"
+#include "components/value_store/value_store_factory.h"
+#include "components/value_store/value_store_factory_impl.h"
+#include "extensions/browser/api/storage/value_store_util.h"
+#include "extensions/browser/extension_file_task_runner.h"
+#include "extensions/common/constants.h"
```
Comment (simonhong) [inline browser/importer/extensions_import_helpers.cc:20]: Can you check `sources.gni` has missing deps for this including?

File: browser/importer/brave_external_process_importer_host.cc [hunk 6]
```diff
@@ -16,10 +17,15 @@
...
 #include "brave/common/importer/importer_constants.h"
+#include "chrome/browser/profiles/profile.h"
+#include "content/public/browser/browser_task_traits.h"
+#include "content/public/browser/browser_thread.h"
 
 #if BUILDFLAG(ENABLE_EXTENSIONS)
+#include "brave/browser/importer/extensions_import_helpers.h"
 #include "chrome/browser/extensions/webstore_install_with_prompt.h"
 #include "chrome/common/extensions/webstore_install_result.h"
+#include "extensions/browser/extension_file_task_runner.h"
```
Comment (simonhong) [inline browser/importer/brave_external_process_importer_host.cc:28]: ditto for missing deps.

File: browser/importer/brave_external_process_importer_host.h [hunk 7]
```diff
@@ -19,21 +23,35 @@ class BraveExternalProcessImporterHost : public ExternalProcessImporterHost {
       delete;
   BraveExternalProcessImporterHost& operator=(
       const BraveExternalProcessImporterHost&) = delete;
+  using MockedInstallCallback =
+      base::RepeatingCallback<void(const std::string&)>;
+  void DoNotLaunchImportForTesting();
```
Comment (simonhong) [inline browser/importer/brave_external_process_importer_host.h:28]: I think this method could go into `private`.

File: browser/importer/brave_external_process_importer_host_unittest.cc [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+#include "brave/components/constants/brave_paths.h"
+#include "chrome/browser/importer/importer_progress_observer.h"
+#include "chrome/common/extensions/webstore_install_result.h"
+#include "chrome/test/base/testing_profile.h"
+#include "components/value_store/test_value_store_factory.h"
+#include "components/value_store/value_store.h"
+#include "content/public/test/browser_task_environment.h"
+#include "extensions/common/constants.h"
+#include "testing/gmock/include/gmock/gmock.h"
+#include "testing/gtest/include/gtest/gtest.h"
```
Comment (simonhong) [inline browser/importer/brave_external_process_importer_host_unittest.cc:30]: ditto for deps

**Review by simonhong — APPROVED** [review]
++ with some nits.

File: browser/importer/brave_external_process_importer_host_unittest.cc (hunk 8 above)
Comment (spylogsster) [reply browser/importer/brave_external_process_importer_host_unittest.cc:30]: done

File: browser/importer/brave_external_process_importer_host.h (hunk 7 above)
Comment (spylogsster) [reply browser/importer/brave_external_process_importer_host.h]: done

File: browser/importer/brave_external_process_importer_host.cc (hunk 6 above)
Comment (spylogsster) [reply browser/importer/brave_external_process_importer_host.cc:28]: done

File: browser/importer/extensions_import_helpers.cc (hunk 5 above)
Comment (spylogsster) [reply browser/importer/extensions_import_helpers.cc:20]: done

**Comment by bsclifton** [comment]
Built and ran locally - this works great; code looks good too 😄 Thanks for the extra work on the tests!

This is basically importing the `Local Extension Settings` content from the source. I notice there is also a `Extension State` folder... looking at the Chromium code, I see that referenced here for example
https://source.chromium.org/chromium/chromium/src/+/main:extensions/browser/state_store.cc;l=85?q=kStateStoreName&ss=chromium

Is this storage of interest also?

**Review by bsclifton — APPROVED** [review]
Left a comment - but this is great! 😄👍

**Comment by spylogsster** [comment]
> Built and ran locally - this works great; code looks good too 😄 Thanks for the extra work on the tests!
> 
> This is basically importing the `Local Extension Settings` content from the source. I notice there is also a `Extension State` folder... looking at the Chromium code, I see that referenced here for example https://source.chromium.org/chromium/chromium/src/+/main:extensions/browser/state_store.cc;l=85?q=kStateStoreName&ss=chromium
> 
> Is this storage of interest also?

it doesnt look like extensions settings that has to be imported, some internal chromium states

**Pull request merged**